File: src/dotnet/ReSharperPlugin.SpecflowRiderPlugin/Psi/GherkinParser.cs
using System.Text.RegularExpressions;
using JetBrains.Annotations;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.ReSharper.Psi;
using JetBrains.ReSharper.Psi.ExtensionsAPI.Tree;
using JetBrains.ReSharper.Psi.Parsing;
using JetBrains.ReSharper.Psi.Tree;
using JetBrains.ReSharper.Psi.TreeBuilder;
using ReSharperPlugin.SpecflowRiderPlugin.Caching.SpecflowJsonSettings;

namespace ReSharperPlugin.SpecflowRiderPlugin.Psi
{
    public class GherkinParser : IParser
    {
        // ReSharper disable once InconsistentNaming
        private static readonly NodeTypeSet SCENARIO_END_TOKENS = new NodeTypeSet(GherkinTokenTypes.BACKGROUND_KEYWORD,
                                                                                  GherkinTokenTypes.SCENARIO_KEYWORD,
                                                                                  GherkinTokenTypes.SCENARIO_OUTLINE_KEYWORD,
                                                                                  GherkinTokenTypes.RULE_KEYWORD,
                                                                                  GherkinTokenTypes.FEATURE_KEYWORD);
        private static readonly Regex LanguagePattern = new Regex ("^\\s*#\\s*language\\s*:\\s*(?<lang>[a-zA-Z\\-_]+)\\s*$", RegexOptions.Compiled);

        private readonly ILexer _lexer;
        private readonly IPsiSourceFile _sourceFile;
        private readonly GherkinKeywordProvider _keywordProvider;
        private GherkinStepKind _lastStepKind = GherkinStepKind.Given;
        [CanBeNull] private string _lang;

        public GherkinParser(ILexer lexer, IPsiSourceFile sourceFile, SpecflowSettingsProvider settingsProvider, GherkinKeywordProvider keywordProvider)
        {
            var settings = settingsProvider.GetDefaultSettings();
            _lang = settings.Language.NeutralFeature;
            _lexer = lexer;
            _sourceFile = sourceFile;
            _keywordProvider = keywordProvider;
        }

        public IFile ParseFile()
        {
            using (var lifetimeDefinition = Lifetime.Define())
            {
                var builder = new PsiBuilder(_lexer, GherkinNodeTypes.FILE, null, lifetimeDefinition.Lifetime);
                var fileMarker = builder.Mark();

                while (!builder.Eof())
                {
                    var tokenType = builder.GetTokenType();

                    if (tokenType == GherkinTokenTypes.FEATURE_KEYWORD)
                        ParseFeature(builder);
                    else if (tokenType == GherkinTokenTypes.TAG)
                        ParseTags(builder);
                    else if (tokenType == GherkinTokenTypes.COMMENT)
                        ParseComments(builder);
                    else
                        builder.AdvanceLexer();
                }

                builder.Done(fileMarker, GherkinNodeTypes.FILE, new GherkinFile.FileMetadata(_sourceFile.Name, _lang));
                var resultTree = (GherkinFile) builder.BuildTree();

                return resultTree;
            }
        }

        private void ParseComments(PsiBuilder builder)
        {
            while (builder.GetTokenType() == GherkinTokenTypes.COMMENT)
            {
                var commentMarker = builder.Mark();
                var commentText = builder.GetTokenText();
                builder.AdvanceLexer();
                var match = LanguagePattern.Match(commentText);
                if (match.Success)
                {
                    _lang = match.Groups["lang"].Value;
                    builder.Done(commentMarker, GherkinNodeTypes.LANGUAGE_COMMENT, _lang);
                }
                else
                    builder.Drop(commentMarker);

                if (builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
                    builder.AdvanceLexer();
            }
        }

        private static void ParseTags(PsiBuilder builder)
        {
            while (builder.GetTokenType() == GherkinTokenTypes.TAG)
            {
                var tagMarker = builder.Mark();
                builder.AdvanceLexer();
                builder.Done(tagMarker, GherkinNodeTypes.TAG, null);

                if (builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
                    builder.AdvanceLexer();
            }
        }

        private void ParseFeature(PsiBuilder builder)
        {
            var featureMarker = builder.Mark();

            Assertion.Assert(builder.GetTokenType() == GherkinTokenTypes.FEATURE_KEYWORD,
                             "_builder.GetTokenType() == GherkinTokenTypes.FEATURE_KEYWORD");

            int? descMarker = null;
            bool wasLineBreak = false;
            do
            {
                builder.AdvanceLexer();

                var tokenType = builder.GetTokenType();
                if (tokenType == GherkinTokenTypes.TEXT && descMarker == null)
                {
                    if (wasLineBreak)
                        descMarker = builder.Mark();
                }

                if (GherkinTokenTypes.SCENARIOS_KEYWORDS[tokenType] ||
                    tokenType == GherkinTokenTypes.RULE_KEYWORD ||
                    tokenType == GherkinTokenTypes.BACKGROUND_KEYWORD ||
                    tokenType == GherkinTokenTypes.TAG)
                {
                    if (descMarker != null)
                    {
                        builder.Done(descMarker.Value, GherkinNodeTypes.FEATURE_HEADER, null);
                        descMarker = null;
                    }

                    ParseFeatureElements(builder);
                }

                wasLineBreak = IsLineBreak(builder);
            } while (builder.GetTokenType() != GherkinTokenTypes.FEATURE_KEYWORD && !builder.Eof());

            if (descMarker != null)
                builder.Done(descMarker.Value, GherkinNodeTypes.FEATURE_HEADER, null);

            builder.Done(featureMarker, GherkinNodeTypes.FEATURE, null);
        }

        private void ParseFeatureElements(PsiBuilder builder)
        {
            int? ruleMarker = null;
            while (builder.GetTokenType() != GherkinTokenTypes.FEATURE_KEYWORD && !builder.Eof())
            {
                if (builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
                {
                    builder.AdvanceLexer();
                    continue;
                }

                ruleMarker = ParseRule(builder, ruleMarker);
                ParseScenario(builder);
            }

            if (ruleMarker.HasValue)
                builder.Done(ruleMarker.Value, GherkinNodeTypes.RULE, null);
        }

        private static int? ParseRule(PsiBuilder builder, int? ruleMarker)
        {
            if (builder.GetTokenType() == GherkinTokenTypes.RULE_KEYWORD)
            {
                if (ruleMarker != null)
                    builder.Done(ruleMarker.Value, GherkinNodeTypes.RULE, null);

                ruleMarker = builder.Mark();
                builder.AdvanceLexer();
                if (builder.GetTokenType() == GherkinTokenTypes.COLON)
                    builder.AdvanceLexer();
                else
                    return ruleMarker;

                while (builder.GetTokenType() == GherkinTokenTypes.TEXT ||
                       builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
                    builder.AdvanceLexer();
            }

            return ruleMarker;
        }

        private void ParseScenario(PsiBuilder builder)
        {
            var scenarioMarker = builder.Mark();
            ParseTags(builder);

            // scenarios
            var startTokenType = builder.GetTokenType();
            var outline = startTokenType == GherkinTokenTypes.SCENARIO_OUTLINE_KEYWORD;
            builder.AdvanceLexer();

            while (!AtScenarioEnd(builder))
            {
                ParseTags(builder);

                if (ParseStepParameter(builder))
                    continue;

                if (builder.GetTokenType() == GherkinTokenTypes.STEP_KEYWORD)
                    ParseStep(builder);
                else if (builder.GetTokenType() == GherkinTokenTypes.EXAMPLES_KEYWORD)
                    ParseExamplesBlock(builder);
                else
                    builder.AdvanceLexer();
            }

            builder.Done(scenarioMarker, outline ? GherkinNodeTypes.SCENARIO_OUTLINE : GherkinNodeTypes.SCENARIO, null);
        }

        private void ParseStep(PsiBuilder builder)
        {
            var marker = builder.Mark();
            var keywordText = builder.GetTokenText();
            builder.AdvanceLexer();
            while (builder.GetTokenType() == GherkinTokenTypes.TEXT ||
                   builder.GetTokenType() == GherkinTokenTypes.STEP_PARAMETER_BRACE ||
                   builder.GetTokenType() == GherkinTokenTypes.STEP_PARAMETER_TEXT ||
                   builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
            {
                if (IsLineBreak(builder))
                    break;

                if (!ParseStepParameter(builder))
                    builder.AdvanceLexer();
            }

            var nextToken = builder.GetTokenType(1);
            if (nextToken == GherkinTokenTypes.PIPE)
            {
                builder.AdvanceLexer();
                ParseTable(builder);
            }
            else if (nextToken == GherkinTokenTypes.PYSTRING)
            {
                builder.AdvanceLexer();
                ParsePystring(builder);
            }

            var stepKind = _keywordProvider.GetStepKind(_lang, keywordText);
            if (stepKind == GherkinStepKind.And)
                stepKind = _lastStepKind;
            else if (stepKind == GherkinStepKind.Given || stepKind == GherkinStepKind.When || stepKind == GherkinStepKind.Then)
                _lastStepKind = stepKind;
            builder.Done(marker, GherkinNodeTypes.STEP, stepKind);
        }

        private static void ParseExamplesBlock(PsiBuilder builder)
        {
            var marker = builder.Mark();
            builder.AdvanceLexer();
            if (builder.GetTokenType() == GherkinTokenTypes.COLON)
                builder.AdvanceLexer();

            while (builder.GetTokenType() == GherkinTokenTypes.TEXT ||
                   builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
                builder.AdvanceLexer();

            if (builder.GetTokenType() == GherkinTokenTypes.PIPE)
                ParseTable(builder);

            builder.Done(marker, GherkinNodeTypes.EXAMPLES_BLOCK, null);
        }

        private static void ParseTable(PsiBuilder builder)
        {
            var marker = builder.Mark();
            var rowMarker = builder.Mark();
            var headerNodeType = GherkinNodeTypes.TABLE_HEADER_ROW;
            int? cellMarker = null;

            var wasLineBreak = false;
            var possibleEmptyCell = false;
            while (builder.GetTokenType() == GherkinTokenTypes.PIPE ||
                   builder.GetTokenType() == GherkinTokenTypes.TABLE_CELL ||
                   builder.GetTokenType() == GherkinTokenTypes.WHITE_SPACE)
            {
                var tokenType = builder.GetTokenType();
                if (tokenType == GherkinTokenTypes.TABLE_CELL && cellMarker == null)
                {
                    cellMarker = builder.Mark();
                }
                else if (tokenType != GherkinTokenTypes.TABLE_CELL && cellMarker != null)
                {
                    builder.Done(cellMarker.Value, GherkinNodeTypes.TABLE_CELL, null);
                    cellMarker = null;
                    possibleEmptyCell = false;
                }

                if (tokenType == GherkinTokenTypes.PIPE)
                {
                    if (wasLineBreak)
                    {
                        possibleEmptyCell = true;
                        builder.Done(rowMarker, headerNodeType, null);
                        headerNodeType = GherkinNodeTypes.TABLE_ROW;
                        rowMarker = builder.Mark();
                    }
                    else
                    {
                        if (possibleEmptyCell)
                        {
                            cellMarker = builder.Mark();
                            builder.Done(cellMarker.Value, GherkinNodeTypes.TABLE_CELL, null);
                            cellMarker = null;
                        }

                        possibleEmptyCell = true;
                    }
                }

                wasLineBreak = IsLineBreak(builder);
                builder.AdvanceLexer();
            }

            if (cellMarker.HasValue)
                builder.Done(cellMarker.Value, GherkinNodeTypes.TABLE_CELL, null);

            builder.Done(rowMarker, headerNodeType, null);
            builder.Done(marker, GherkinNodeTypes.TABLE, null);
        }

        private static void ParsePystring(PsiBuilder builder)
        {
            if (!builder.Eof())
            {
                var marker = builder.Mark();
                builder.AdvanceLexer();
                while (!builder.Eof() && builder.GetTokenType() != GherkinTokenTypes.PYSTRING)
                {
                    if (!ParseStepParameter(builder))
                        builder.AdvanceLexer();
                }

                if (!builder.Eof())
                    builder.AdvanceLexer();

                builder.Done(marker, GherkinNodeTypes.PYSTRING, null);
            }
        }

        private static bool ParseStepParameter(PsiBuilder builder)
        {
            if (builder.GetTokenType() != GherkinTokenTypes.STEP_PARAMETER_TEXT)
                return false;

            var stepParameterMarker = builder.Mark();
            builder.AdvanceLexer();
            builder.Done(stepParameterMarker, GherkinNodeTypes.STEP_PARAMETER, null);
            return true;
        }

        private static bool AtScenarioEnd(PsiBuilder builder)
        {
            var i = 0;
            while (builder.GetTokenType(i) == GherkinTokenTypes.TAG ||
                   builder.GetTokenType(i) == GherkinTokenTypes.WHITE_SPACE)
                i++;

            var tokenType = builder.GetTokenType(i);
            return tokenType == null || SCENARIO_END_TOKENS[tokenType];
        }

        private static bool IsLineBreak(PsiBuilder builder)
        {
            if (builder.GetTokenType() != GherkinTokenTypes.WHITE_SPACE)
                return false;

            return builder.GetTokenText()?.Contains("\n") == true;
        }
    }
}

File: src/rider/main/kotlin/com/jetbrains/rider/plugins/reqnrollriderplugin/psi/GherkinParser.kt
package com.jetbrains.rider.plugins.reqnrollriderplugin.psi

import com.intellij.lang.*
import com.intellij.psi.tree.IElementType
import com.intellij.psi.tree.TokenSet

class GherkinParser : PsiParser {
    override fun parse(root: IElementType, builder: PsiBuilder): ASTNode {
        val marker = builder.mark()
        parseFileTopLevel(builder)
        marker.done(GherkinParserDefinition.GHERKIN_FILE)
        return builder.treeBuilt
    }

    companion object {
        private val SCENARIO_END_TOKENS = TokenSet.create(
            GherkinTokenTypes.BACKGROUND_KEYWORD,
            GherkinTokenTypes.SCENARIO_KEYWORD,
            GherkinTokenTypes.SCENARIO_OUTLINE_KEYWORD,
            GherkinTokenTypes.RULE_KEYWORD,
            GherkinTokenTypes.FEATURE_KEYWORD
        )

        private fun parseFileTopLevel(builder: PsiBuilder) {
            while (!builder.eof()) {
                val tokenType = builder.tokenType
                if (tokenType === GherkinTokenTypes.FEATURE_KEYWORD) {
                    parseFeature(builder)
                } else if (tokenType === GherkinTokenTypes.TAG) {
                    parseTags(builder)
                } else {
                    builder.advanceLexer()
                }
            }
        }

        private fun parseFeature(builder: PsiBuilder) {
            val marker = builder.mark()
            assert(builder.tokenType === GherkinTokenTypes.FEATURE_KEYWORD)
            val featureEnd = builder.currentOffset + getTokenLength(builder.tokenText)
            var descMarker: PsiBuilder.Marker? = null
            while (true) {
                val tokenType = builder.tokenType
                if (tokenType === GherkinTokenTypes.TEXT && descMarker == null) {
                    if (hadLineBreakBefore(builder, featureEnd)) {
                        descMarker = builder.mark()
                    }
                }
                if (GherkinTokenTypes.SCENARIOS_KEYWORDS.contains(tokenType) || tokenType === GherkinTokenTypes.RULE_KEYWORD || tokenType === GherkinTokenTypes.BACKGROUND_KEYWORD || tokenType === GherkinTokenTypes.TAG) {
                    if (descMarker != null) {
                        descMarker.done(GherkinElementTypes.FEATURE_HEADER)
                        descMarker = null
                    }
                    parseFeatureElements(builder)
                    if (builder.tokenType === GherkinTokenTypes.FEATURE_KEYWORD) {
                        break
                    }
                }
                builder.advanceLexer()
                if (builder.eof()) break
            }
            descMarker?.done(GherkinElementTypes.FEATURE_HEADER)
            marker.done(GherkinElementTypes.FEATURE)
        }

        private fun hadLineBreakBefore(builder: PsiBuilder, prevTokenEnd: Int): Boolean {
            if (prevTokenEnd < 0) return false
            val precedingText = builder.originalText.subSequence(prevTokenEnd, builder.currentOffset).toString()
            return precedingText.contains("\n")
        }

        private fun parseTags(builder: PsiBuilder) {
            while (builder.tokenType === GherkinTokenTypes.TAG) {
                val tagMarker = builder.mark()
                builder.advanceLexer()
                tagMarker.done(GherkinElementTypes.TAG)
            }
        }

        private fun parseFeatureElements(builder: PsiBuilder) {
            var ruleMarker: PsiBuilder.Marker? = null
            while (builder.tokenType !== GherkinTokenTypes.FEATURE_KEYWORD && !builder.eof()) {
                if (builder.tokenType === GherkinTokenTypes.RULE_KEYWORD) {
                    ruleMarker?.done(GherkinElementTypes.RULE)
                    ruleMarker = builder.mark()
                    builder.advanceLexer()
                    if (builder.tokenType === GherkinTokenTypes.COLON) {
                        builder.advanceLexer()
                    } else {
                        break
                    }
                    while (builder.tokenType === GherkinTokenTypes.TEXT) {
                        builder.advanceLexer()
                    }
                }
                val marker = builder.mark()
                // tags
                parseTags(builder)

                // scenarios
                val startTokenType = builder.tokenType
                val outline = startTokenType === GherkinTokenTypes.SCENARIO_OUTLINE_KEYWORD
                builder.advanceLexer()
                parseScenario(builder)
                marker.done(if (outline) GherkinElementTypes.SCENARIO_OUTLINE else GherkinElementTypes.SCENARIO)
            }
            ruleMarker?.done(GherkinElementTypes.RULE)
        }

        private fun parseScenario(builder: PsiBuilder) {
            while (!atScenarioEnd(builder)) {
                if (builder.tokenType === GherkinTokenTypes.TAG) {
                    val marker = builder.mark()
                    parseTags(builder)
                    if (atScenarioEnd(builder)) {
                        marker.rollbackTo()
                        break
                    } else {
                        marker.drop()
                    }
                }
                if (parseStepParameter(builder)) {
                    continue
                }
                if (builder.tokenType === GherkinTokenTypes.STEP_KEYWORD) {
                    parseStep(builder)
                } else if (builder.tokenType === GherkinTokenTypes.EXAMPLES_KEYWORD) {
                    parseExamplesBlock(builder)
                } else {
                    builder.advanceLexer()
                }
            }
        }

        private fun atScenarioEnd(builder: PsiBuilder): Boolean {
            var i = 0
            while (builder.lookAhead(i) === GherkinTokenTypes.TAG) {
                i++
            }
            val tokenType = builder.lookAhead(i)
            return tokenType == null || SCENARIO_END_TOKENS.contains(tokenType)
        }

        private fun parseStepParameter(builder: PsiBuilder): Boolean {
            if (builder.tokenType === GherkinTokenTypes.STEP_PARAMETER_TEXT) {
                val stepParameterMarker = builder.mark()
                builder.advanceLexer()
                stepParameterMarker.done(GherkinElementTypes.STEP_PARAMETER)
                return true
            }
            return false
        }

        private fun parseStep(builder: PsiBuilder) {
            val marker = builder.mark()
            builder.advanceLexer()
            var prevTokenEnd = -1
            while (builder.tokenType === GherkinTokenTypes.TEXT || builder.tokenType === GherkinTokenTypes.STEP_PARAMETER_BRACE || builder.tokenType === GherkinTokenTypes.STEP_PARAMETER_TEXT) {
                val tokenText = builder.tokenText
                if (hadLineBreakBefore(builder, prevTokenEnd)) {
                    break
                }
                prevTokenEnd = builder.currentOffset + getTokenLength(tokenText)
                if (!parseStepParameter(builder)) {
                    builder.advanceLexer()
                }
            }
            val tokenTypeAfterName = builder.tokenType
            if (tokenTypeAfterName === GherkinTokenTypes.PIPE) {
                parseTable(builder)
            } else if (tokenTypeAfterName === GherkinTokenTypes.PYSTRING) {
                parsePystring(builder)
            }
            marker.done(GherkinElementTypes.STEP)
        }

        private fun parsePystring(builder: PsiBuilder) {
            if (!builder.eof()) {
                val marker = builder.mark()
                builder.advanceLexer()
                while (!builder.eof() && builder.tokenType !== GherkinTokenTypes.PYSTRING) {
                    if (!parseStepParameter(builder)) {
                        builder.advanceLexer()
                    }
                }
                if (!builder.eof()) {
                    builder.advanceLexer()
                }
                marker.done(GherkinElementTypes.PYSTRING)
            }
        }

        private fun parseExamplesBlock(builder: PsiBuilder) {
            val marker = builder.mark()
            builder.advanceLexer()
            if (builder.tokenType === GherkinTokenTypes.COLON) builder.advanceLexer()
            while (builder.tokenType === GherkinTokenTypes.TEXT) {
                builder.advanceLexer()
            }
            if (builder.tokenType === GherkinTokenTypes.PIPE) {
                parseTable(builder)
            }
            marker.done(GherkinElementTypes.EXAMPLES_BLOCK)
        }

        private fun parseTable(builder: PsiBuilder) {
            val marker = builder.mark()
            var rowMarker = builder.mark()
            var prevCellEnd = -1
            var isHeaderRow = true
            var cellMarker: PsiBuilder.Marker? = null
            var prevToken: IElementType? = null
            while (builder.tokenType === GherkinTokenTypes.PIPE || builder.tokenType === GherkinTokenTypes.TABLE_CELL) {
                val tokenType = builder.tokenType
                val hasLineBreakBefore = hadLineBreakBefore(builder, prevCellEnd)

                // cell - is all between pipes
                if (prevToken === GherkinTokenTypes.PIPE) {
                    // Don't start new cell if prev was last in the row
                    // it's not a cell, we just need to close a row
                    if (!hasLineBreakBefore) {
                        cellMarker = builder.mark()
                    }
                }
                if (tokenType === GherkinTokenTypes.PIPE) {
                    if (cellMarker != null) {
                        closeCell(cellMarker)
                        cellMarker = null
                    }
                }
                if (hasLineBreakBefore) {
                    closeRowMarker(rowMarker, isHeaderRow)
                    isHeaderRow = false
                    rowMarker = builder.mark()
                }
                prevCellEnd = builder.currentOffset + getTokenLength(builder.tokenText)
                prevToken = tokenType
                builder.advanceLexer()
            }
            if (cellMarker != null) {
                closeCell(cellMarker)
            }
            closeRowMarker(rowMarker, isHeaderRow)
            marker.done(GherkinElementTypes.TABLE)
        }

        private fun closeCell(cellMarker: PsiBuilder.Marker) {
            cellMarker.done(GherkinElementTypes.TABLE_CELL)
        }

        private fun closeRowMarker(rowMarker: PsiBuilder.Marker, headerRow: Boolean) {
            rowMarker.done(if (headerRow) GherkinElementTypes.TABLE_HEADER_ROW else GherkinElementTypes.TABLE_ROW)
        }

        private fun getTokenLength(tokenText: String?): Int {
            return tokenText?.length ?: 0
        }
    }
}
